base/broadcast.jl

```diff
@@ -52,10 +52,27 @@ BroadcastStyle(::Type{Union{}}) = Unknown()  # ambiguity resolution
 """
 `Broadcast.Scalar()` is a [`BroadcastStyle`](@ref) indicating that an object is not
 treated as a container for the purposes of broadcasting. This is the default for objects
-that have not customized `BroadcastStyle`.
+that have neither customized `BroadcastStyle` nor implemented the [`start`](@ref) method
+(for iterator types).
 """
 struct Scalar <: BroadcastStyle end
-BroadcastStyle(::Type) = Scalar()
+hasshape_ndims(::Base.HasShape{N}) where {N} = N
+function BroadcastStyle(::Type{T}) where T
+    if method_exists(start, Tuple{T})
```

Comment:
This is not inferrable, and thus might be pretty bad for code like
```julia
function foo(x)
    y = Float64.(x)
    # now do something with y
end
```
Unfortunately #16422 wouldn't help.

At the same time, I recognize that any problematic type can be optimized by adding a specific defintion.

At a minimum we may have to change the `typeof` calls in `collect_styles` to `Core.Typeof`, and then specialize
```julia
BroadcastStyle(::Type{Type{T}}) where T = Scalar()
```

Reply:
We could require iterators to define a trait. That would be more consistent with what we do elsewhere, and that wouldn't be a terrible burden either. I had contemplated adding an `NotIterable` type to `Base.iteratorsize`, but that could also be a separate function like `isiterable`.

Reply:
On balance I think it's better to require iterators to define a trait, and make scalars the default.

One slightly-crazy thought is that the trait name could be the output of `BroadcastStyle`. But on balance I'm not sure this is a good idea, because there may be reasons to have things that act like scalars that don't return `Scalar()`. It's probably better to have a separate `isiterable` trait.

Reply:
@vtjnash mentioned [here](https://github.com/mauro3/SimpleTraits.jl/issues/40#issuecomment-293633610) that `method_exists` could be made inferable now, presumably circumnavigating the problem mentioned in #16422.

Reply:
It could, but it hasn't been in the past since we don't want the coupling. Also, #25261 will break this.